mlfinlab/data_structures/imbalance_data_structures.py

```diff
@@ -0,0 +1,287 @@
+"""
+Advances in Financial Machine Learning, Marcos Lopez de Prado
+Chapter 2: Financial Data Structures
+
+This module contains the functions to help users create structured financial data from raw unstructured data,
+in the form of tick, volume, and dollar imbalance bars.
+
+These bars are used throughout the text book (Advances in Financial Machine Learning, By Marcos Lopez de Prado, 2018, pg 25)
+to build the more interesting features for predicting financial time series data.
+
+These financial data structures have better statistical properties when compared to those based on fixed time interval sampling.
+A great paper to read more about this is titled: The Volume Clock: Insights into the high frequency paradigm, Lopez de Prado, et al
+"""
+
+# Imports
+import pandas as pd
+import numpy as np
+from mlfinlab.data_structures.fast_ewma import ewma
+
+
+def _update_counters(cache, flag, exp_num_ticks_init):
+    """
+    Updates the counters by resetting them or making use of the cache to update them based on a previous batch.
+
+    :param cache: Contains information from the previous batch that is relevant in this batch.
+    :param flag: A flag which signals to use the cache.
+    :param exp_num_ticks: Expected number of ticks per bar
+    :param ewma_window: ewma_window to estimate imbalance
+    :return: Updated counters - cum_ticks, cum_dollar_value, cum_volume, high_price, low_price, exp_num_ticks, imbalance_array
+    """
+    # Check flag
+    if flag and cache:
+        # Update variables based on cache
+        cum_ticks = int(cache[-1][6])
```

Comment:
It's too dangerous to use simple indexing with cache. Consider using something like namedtuple. It will add more intuition and simplify handling